data/transactions/logic/eval.go

```diff
@@ -649,10 +703,68 @@ func EvalContract(program []byte, gi int, aid basics.AppIndex, params *EvalParam
 		}
 	}
 
+	// If this is a creation, make any "0 index" box refs available now that we
+	// have an appID.
+	if cx.txn.Txn.ApplicationID == 0 {
+		for _, br := range cx.txn.Txn.Boxes {
+			if br.Index == 0 {
+				cx.EvalParams.available.boxes[boxRef{cx.appID, string(br.Name)}] = boxTouch{}
+			}
+		}
+	}
+
+	// Check the I/O budget for reading if this is the first top-level app call
+	if cx.caller == nil && !cx.readBudgetChecked {
+		boxRefCount := uint64(0) // Intentionally counts duplicates
+		for _, tx := range cx.TxnGroup {
+			boxRefCount += uint64(len(tx.Txn.Boxes))
+		}
+		cx.ioBudget = boxRefCount * cx.Proto.BytesPerBoxReference
+
+		used := uint64(0)
+		for br := range cx.available.boxes {
+			if len(br.name) == 0 {
+				// 0 length names are not allowed for actual created boxes, but
+				// may have bene used to add I/O budget.
+				continue
+			}
+			box, ok, err := cx.Ledger.GetBox(br.app, br.name)
+			if err != nil {
+				return false, nil, err
+			}
+			if !ok {
+				continue
+			}
+			size := uint64(len(box))
+			cx.available.boxes[br] = boxTouch{false, size}
+
+			used = basics.AddSaturate(used, size)
+			if used > cx.ioBudget {
+				return false, nil, fmt.Errorf("box read budget (%d) exceeded", cx.ioBudget)
+			}
+		}
+		cx.readBudgetChecked = true
+	}
+
 	if cx.Trace != nil && cx.caller != nil {
 		fmt.Fprintf(cx.Trace, "--- enter %d %s %v\n", aid, cx.txn.Txn.OnCompletion, cx.txn.Txn.ApplicationArgs)
 	}
 	pass, err := eval(program, &cx)
+
+	// Check the I/O budget for writing. TODO: Only need to do this once if we
+	// figure out we're the last app call in the group.
+	if cx.caller == nil {
+		used := uint64(0)
+		for _, bt := range cx.available.boxes {
+			if bt.dirty {
+				used = basics.AddSaturate(used, bt.size)
+			}
+			if used > cx.ioBudget {
```

Comment:
nit: doesn't really matter, but this code could be placed in the above if statement, since that's the only time `used` is modified